fontbe/src/kern.rs

```diff
@@ -124,29 +210,72 @@ impl Work<Context, AnyWorkId, Error> for KerningWork {
                             as u16,
                     );
                     let right = glyph_classes
-                        .get(&right)
+                        .get(right)
                         .ok_or_else(|| Error::MissingGlyphId(right.clone()))?;
                     for gid1 in right.iter() {
-                        builder.insert_pair(gid0, x_adv_record.clone(), gid1, empty.clone());
+                        kerns.push(PairPosEntry::Pair(
+                            gid0,
+                            x_adv_record.clone(),
+                            gid1,
+                            empty.clone(),
+                        ));
                     }
                 }
                 (KernParticipant::Group(left), KernParticipant::Glyph(right)) => {
                     let left = glyph_classes
                         .get(left)
                         .ok_or_else(|| Error::MissingGlyphId(left.clone()))?;
-                    let gid1 = gid(right)?;
+                    let gid1 = gid(glyph_order, right)?;
                     for gid0 in left.iter() {
-                        builder.insert_pair(gid0, x_adv_record.clone(), gid1, empty.clone());
+                        kerns.push(PairPosEntry::Pair(
+                            gid0,
+                            x_adv_record.clone(),
+                            gid1,
+                            empty.clone(),
+                        ));
                     }
                 }
             }
         }
 
-        let mut kerning = Kerning::default();
+        context.kern_segments.set(KernSegment {
+            segment: self.segment,
+            kerns,
+        });
+
+        Ok(())
+    }
+}
+
+impl Work<Context, AnyWorkId, Error> for KerningGatherWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Kerns.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        AccessBuilder::new()
+            .variant(WorkId::KernPairs) // wait for pairs because kern segments don't spawn until pairs are done
+            .variant(WorkId::KernSegment(0))
+            .build()
+    }
+
+    fn exec(&self, context: &Context) -> Result<(), Error> {
+        let segments = context.kern_segments.all();
+        let mut builder = PairPosBuilder::default();
+
+        for ppe in segments
+            .iter()
+            .flat_map(|(_, segment)| segment.kerns.iter())
+            .cloned()
+        {
+            ppe.add_to(&mut builder);
+        }
```

Comment:
(how) do we ensure that the pairs are added to the builder in a deterministic order? IIUC, the kern_segments are produced in different threads in parallel so we might not guarantee their ordering is the same as if they were built serially

Reply:
Or perhaps the order does not matter because the builder will sort them somehow?

Reply:
I tried to compile GS multiple times from this branch, and then diff the resulting ttx files, and can see that GDEF and GPOS tables are different each time I run fontc.. Maybe we should have a CI check that ensures that building the same source file multiple times produces identical results.

Reply:
actually this happens from the current main branch too.. Something very weird is going on, the size of the GPOS table seems to fluctuate, and ttx dumps contain more or less data every time I run fontc on the same GoogleSans.designspace...

Reply:
it seems like the indeterminism of GDEF/GPOS has been introduced after https://github.com/googlefonts/fontc/pull/642 got merged to main, if I checkout immediately before that it seems produce identical GDEF and GPOS tables each time I re-run fontc.

Reply:
besides GDEF/GPOS, the files are still not 100% identical. I see diffs in GSUB with stylistic set features using FeatureParams, the nameID seems to change every time I rebuild.. Ok time to file a separate issue

```diff
--- /dev/fd/63  2023-12-13 13:24:16.601441808 +0000
+++ /dev/fd/62  2023-12-13 13:24:16.601441808 +0000
@@ -7635,7 +7635,7 @@
     <flags value="00000000 00000011"/>
     <unitsPerEm value="1000"/>
     <created value="Thu Jul  6 09:41:44 2017"/>
-    <modified value="Wed Dec 13 13:18:42 2023"/>
+    <modified value="Wed Dec 13 13:20:01 2023"/>
     <xMin value="-978"/>
     <yMin value="-989"/>
     <xMax value="3284"/>
@@ -4137695,7 +4137695,7 @@
         <Feature>
           <FeatureParamsStylisticSet>
             <Version value="0"/>
-            <UINameID value="258"/>  <!-- Grade -->
+            <UINameID value="263"/>  <!-- Text Medium -->
           </FeatureParamsStylisticSet>
           <!-- LookupCount=1 -->
           <LookupListIndex index="0" value="1462"/>
@@ -4137706,7 +4137706,7 @@
         <Feature>
           <FeatureParamsStylisticSet>
             <Version value="0"/>
-            <UINameID value="257"/>  <!-- Weight -->
+            <UINameID value="261"/>  <!-- Bold -->
           </FeatureParamsStylisticSet>
           <!-- LookupCount=1 -->
           <LookupListIndex index="0" value="1463"/>
@@ -4137728,7 +4137728,7 @@
         <Feature>
           <FeatureParamsStylisticSet>
             <Version value="0"/>
-            <UINameID value="260"/>  <!-- Medium -->
+            <UINameID value="256"/>  <!-- Size -->
           </FeatureParamsStylisticSet>
           <!-- LookupCount=1 -->
           <LookupListIndex index="0" value="1464"/>
```

Reply:
https://github.com/googlefonts/fontc/issues/647

Reply:
it should be easy enough to track the start position of each segment, and then sort them here at the end?

Reply:
Fantastic catch, we're clearly missing a test :( Will address in a subsequent PR.